README.md
# distromax
![](https://github.com/Rodrigo-Tenorio/distromax/blob/dc4ba4cf996fde5b51dc23a2611ce347e46c24c0/logo/distromax_logo.png)
[![arXiv](https://img.shields.io/badge/arXiv-2111.12032-b31b1b.svg)](https://arxiv.org/abs/2111.12032)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5763765.svg)](https://doi.org/10.5281/zenodo.5763765)
[![Anaconda-Server Badge](https://anaconda.org/conda-forge/distromax/badges/license.svg)](https://anaconda.org/conda-forge/distromax)
[![Integration Tests](https://github.com/Rodrigo-Tenorio/distromax/actions/workflows/tests.yml/badge.svg)](https://github.com/Rodrigo-Tenorio/distromax/actions/workflows/tests.yml)
[![Anaconda-Server Badge](https://anaconda.org/conda-forge/distromax/badges/version.svg)](https://anaconda.org/conda-forge/distromax)
[![PyPI version](https://badge.fury.io/py/distromax.svg)](https://badge.fury.io/py/distromax)

Empirically estimating the distribution of the loudest candidate from a gravitational-wave search

This package implements the methods described in [Tenorio, Modafferi, Keitel, Sintes (2021)](https://arxiv.org/abs/2111.12032)
to estimate the distribution of the loudest candidate from a search. 

The actual implementation includes:

- [fit.py](distromax/fit.py):
    - `BatchMaxGumbel`: Basic `distromax` method. Construct the batchmax distribution of a set of samples and 
    return the corresponding Gumbel fit. Max. propagation is included as a method.
    - `BatchMaxGumbelNotchingOutlier`: Thin wrapper around `BatchMaxGumbel` to notch narrow-band
    outliers before computing the batchmax distribution. The specifics of this implementation are discussed
    on appendix B of the accompanying publication. 
- [analytical.py](distromax/analytical.py):
    - `AnalyticalGammaToGumbel`: Compute the Gumbel distribution associated to the maximum of
    a set of independent Gamma random variables using the formulas derived 
    in [Gasull, López-Salcedo, Utzet (2015)](https://link.springer.com/article/10.1007%2Fs11749-015-0431-9).

See the [examples](examples) for concrete applications of these classes.

## Citing this work

If `distromax` was useful for your work, we would appreciate if you cite both the
software version DOI under [Zenodo](https://doi.org/10.5281/zenodo.5763765) 
and one or more of the following scientific papers:

- Introduction of `distromax` and description of the method: [Tenorio, Modafferi, Keitel, Sintes, (2021)](https://journals.aps.org/prd/abstract/10.1103/PhysRevD.105.044029) 
  ([inspire](https://inspirehep.net/literature/1974174) / [nasa](https://ui.adsabs.harvard.edu/abs/2021arXiv211112032T/abstract))
- Analytical limit of a Gamma random variable to a Gumbel distribution: 
[A. Gasull, J. López-Salcedo, F. Utzet TEST volume 24, pages 714–733 (2015)](https://link.springer.com/article/10.1007%2Fs11749-015-0431-9)

Here is a better-formatted bibtex entry for the version-independent Zenodo:
```
@misc{distromax,
  author       = {Tenorio, Rodrigo and
                  Modafferi, Luana M. and
                  Keitel, David and
                  Sintes, Alicia M.},
  title        = {distromax: Empirically estimating the distribution of the loudest candidate from a gravitational-wave search},
  month        = dec,
  year         = 2021,
  publisher    = {Zenodo},
  doi          = {10.5281/zenodo.5763765},
  url          = {https://doi.org/10.5281/zenodo.5763765},
  note         = {\url{https://doi.org/10.5281/zenodo.5763765}}
}
```
For individual version DOIs, see the right sidebar at [Zenodo](https://doi.org/10.5281/zenodo.5763765)

## How to install

Please, make sure you are running on a 
[virtual environment](https://docs.python.org/3/library/venv.html) to avoid
any conflicts with system libraries.

The simplest way is to install `distromax` from [PyPI](https://pypi.org/project/distromax/) using `pip`:
```
pip install distromax
```

`distromax` can also be installed using `conda` from the [conda-forge](https://conda-forge.org/#about)
channle as follows:
```
conda install -c conda-forge distromax
```
See [the official documentation](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html)
to learn about conda environments.

To install directly from source, clone this repo with `git clone` and install using `pip`
```
git clone https://github.com/Rodrigo-Tenorio/distromax.git
cd distromax
pip install .
```

## Troubleshooting

In some conservative systems the default `setuptools` may not be the latest version and
installing from source may produce an erorr message such as
```
ERROR: setuptools==44.1.1 is used in combination with setuptools_scm>=6.x

Your build configuration is incomplete and previously worked by accident!


This happens as setuptools is unable to replace itself when a activated build dependency
requires a more recent setuptools version
(it does not respect "setuptools>X" in setup_requires).


setuptools>=31 is required for setup.cfg metadata support
setuptools>=42 is required for pyproject.toml configuration support

Suggested workarounds if applicable:
 - preinstalling build dependencies like setuptools_scm before running setup.py
 - installing setuptools_scm using the system package manager to ensure consistency
 - migrating from the deprecated setup_requires mechanism to pep517/518
   and using a pyproject.toml to declare build dependencies
   which are reliably pre-installed before running the build tools
```

A simple workaround in that case is to update the `pip` and `setuptools` packages
```
pip install --upgrade pip setuptools
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [v1.1.0]

### Added 

- Support Python 3.8, 3.9, 3.10
- Updated journal reference.
- Added Package logo.
- Renamed test scripts to ease pytest's discovery.
- Cleaned up test suite.
- Simplified test workflows.
- Removed `text.usetex` from examples.
- Removed `versioneer` in favour of `setuptools_scm`.
- Added `logging` module.
- Use `pyproject.toml` keeping a minimal `setup.py` for local installs.

### Fixed


distromax/analytical.py
import logging

import numpy as np
from scipy.special import gamma

logger = logging.getLogger(__name__)

class AnalyticalGammaToGumbel:
    """
    Compute relevant quantities related to the Gumbel distribution arising from the
    maximum over an ensemble of iid Gamma-distributed random variables (which includes
    the chi-squared distribution as a special case).
    """

    def __init__(self, dofs=None, shape=None, scale=None, rate=None):
        """
        The single-variable probability distribution can either be described as a
        chi-squared distribution with `dofs` degrees of freedom,
        a Gamma distribution with shape and scale parameters (`shape`, `scale`),
        or a Gamma distribution with shape and rate parameters (`shape`, `rate`).

        A chi-squared distribution with `dofs` degrees of freedom corresponds to a
        Gamma distribution with `shape=dofs/2` and `scale=2`, `rate=1/2`. 

        Parameters
        ----------
        dofs: float, optional
            Degrees of freedom of the underlying chi-squared distribution.
            Authomatically implies `shape=dofs/2` and `scale=2`.
            This parameter is incompatible with `shape`, `scale`, and `rate`.
        shape: float, optional
            Shape parameter of the underlying Gamma distribution.
            This parameter must be given jointly with `scale` or `rate`
            and is incompatible with `dofs`.
        scale: float, optional
            Scale parameter of the underlying gamma distribution 
            (inverse of the rate parameter).
            This parameter must be given jointly with `shape` and is
            incompatible with `dofs` or `rate`.
        rate: float, optional
            Rate parameter of the underlying gamma distribution 
            (inverse of the scale parameter).
            This parameter must be given jointly with `shape` and is
            incompatible with `dofs` or `scale`.
        """
        
        # Check the validity of the input
        if dofs:
            if shape:
                raise ValueError("Only one out of dofs and shape can be given")
            shape = dofs / 2
            scale = 2.
        elif shape:
            if not (bool(scale) ^ bool(rate)):
                raise ValueError("Only one out of scale and rate can be given")
            elif rate:
                scale = 1 / rate
        else:
            raise ValueError("Invalid input parameters: "
                    "Please enter dofs, (scale, shape) or (scale, rate)")
    
        self._set_shape_scale(shape=shape, scale=scale)

    def _set_shape_scale(self, shape, scale):
        self.shape = shape
        self.scale = scale

    def get_gumbel_loc_scale(self, n, improved=True):
        """
        Compute the theoretical estimation of the location and scale parameters of the
        Gumbel distribution resulting from maximizing over a batch of `n` independent
        random variables following the distribution specified by this class.

        We provide two implementations of said parameters, selected by
        switching the `improved` flag.

        If improved is falsy, location and scale parameters are taken from table 3.4.4
        in "Modelling Extremal Events for Insurance and Finance", P. Embrechts,
        C. Klüppelberg, T. Mikosch (1997) https://doi.org/10.1007/978-3-642-33483-2.

        If improved is truthy (recommended), location and scale parameters are taken
        from Eqs. (32) and (33) in  Gasull, López-Salcedo, and Utzet, TEST 24, 714–733
        (2015) https://doi.org/10.1007/s11749-015-0431-9 .

        The improved option returns better estimates for n < 1e6 (and possibly beyond).

        Parameters
        ----------
        n: int
            Number of random variables within the batch being maximized.
        improved: bool
            If True (recommended), apply the improved loc and scale parameters. See
            the main docstring for details.

        Returns
        -------+
        loc: float
            Location parameter of the resulting Gumbel distribution.
        scale: float
            Scale parameter of the resulting Gumbel distribution.
        """

        logn = np.log(n)
        gamma_shape = gamma(self.shape)

        if improved:
            logger.debug("Computing ``improved'' location and scale parameters...")
            logn_gamma = np.log(n / gamma_shape)
            shape_m1 = self.shape - 1
            Bn = logn_gamma + shape_m1 * np.log(shape_m1)
            logBn = np.log(Bn)

            loc = self.scale * (
                logn_gamma
                + shape_m1 * logBn
                + (shape_m1 * shape_m1 * (logBn - np.log(shape_m1)) + shape_m1) / Bn
            )
            scale = (self.scale * loc * (self.scale * shape_m1 + loc)) / (
                loc * loc - self.scale * self.scale * shape_m1 * (self.shape - 2)
            )

        else:
            logger.debug("Computing standard location and scale parameters...")
            loc = self.scale * (
                logn + (self.shape - 1) * np.log(logn) + np.log(gamma_shape)
            )
            scale = self.scale

        return loc, scale

    def get_gumbel_mean_std(self, n, improved=True):
        """
        Compute the theoretical estimation of the mean and standard deviation parameters
        of the Gumbel distribution resulting from maximizing over a batch of `n` independent
        random variables following the distribution specified by this class.

        These results computed from the `loc` and `scale` parameters returned by
        `self.gumbel_mean_std`. See the docstring therein for a further explanation of the
        parameters.

        Parameters
        ----------
        n: int
            Number of random variables within the batch being maximized.
        improved: bool
            If True (recommended), apply the improved loc and scale parameters. See
            the main docstring for details.

        Returns
        -------
        mean: float
            Mean of the Gumbel distribution, computed as `loc + euler_gamma * scale`,
            where `euler_gamma=0.5772...` represents the Euler-Mascheroni constant
            (sequence A001620 in the OEIS).
        std: float
            Standard deviation of the Gumbel distribution,
            computed as `pi/sqrt(6) * scale`.
        """
        loc, scale = self.get_gumbel_loc_scale(n=n, improved=improved)
        mean = loc + np.euler_gamma * scale
        std = np.pi / np.sqrt(6) * scale
        return mean, std
